exonum/src/encoding/fields.rs

```diff
@@ -459,6 +460,56 @@ impl<'a> Field<'a> for SocketAddr {
     }
 }
 
+impl<'a> Field<'a> for ConnectInfo {
+    fn field_size() -> u32 {
+        SocketAddr::field_size() + PUBLIC_KEY_LENGTH as u32
+    }
+
+    unsafe fn read(buffer: &'a [u8], from: u32, to: u32) -> Self {
+        let public_key_offset = from + SocketAddr::field_size();
+
+        let address = SocketAddr::read(buffer, from, public_key_offset);
+
+        debug_assert_eq!(to - public_key_offset, PUBLIC_KEY_LENGTH as u32);
+        let public_key = PublicKey::from_slice(&buffer[public_key_offset as usize..to as usize])
+            .expect("Unable to create PublicKey");
+
+        ConnectInfo {
+            address,
+            public_key,
+        }
+    }
+
+    fn write(&self, buffer: &mut Vec<u8>, from: u32, to: u32) {
+        let public_key_offset = from + SocketAddr::field_size();
+        self.address.write(buffer, from, public_key_offset);
+
+        buffer[public_key_offset as usize..to as usize].copy_from_slice(&self.public_key.as_ref());
+    }
+
+    fn check(
+        buffer: &'a [u8],
+        from: CheckedOffset,
+        to: CheckedOffset,
+        latest_segment: CheckedOffset,
+    ) -> Result {
+        debug_assert_eq!((to - from)?.unchecked_offset(), Self::field_size());
+
+        let public_key_offset =
+            CheckedOffset::new(from.unchecked_offset() + SocketAddr::field_size());
+        let public_key_length = (to - public_key_offset)?.unchecked_offset() as usize;
+
+        if public_key_length != PUBLIC_KEY_LENGTH {
+            return Err(Error::IncorrectPublicKeyLength {
+                expected_length: PUBLIC_KEY_LENGTH,
+                actual_length: public_key_length,
+            });
+        }
```

Comment:
Is length check alone enough to guarantee that `PublicKey::from_slice()` in `read()` never fails? Can `PublicKey` work with literally any byte sequence if it has expected length?

Reply:
Yes, `sodiumoxide` implementation allows creating `PublicKey` from any byte sequence.